### PR TITLE
ARROW-15582: [C++] Add support for registering standard Substrait functions

### PR DESCRIPTION
**Pull request by westonpace**
This picks up where #13285 has left off.  It mostly focuses on the Substrait->Arrow direction at the moment.  In addition, basic support is added for named tables.  This makes it possible to create unit tests that read from in-memory tables instead of requiring unit tests to do a scan.

The PR creates some utilities in `test_plan_builder.h` which allow for the construction of simple Substrait plans programmatically.  This is used to create unit tests for the function mapping.

The PR extracts id "ownership" out of the `ExtensionIdRegistry` and into its own `IdStorage` class.

The PR gets rid of `NestedExtensionIdRegistryImpl` and instead makes `ExtensionIdRegistryImpl` nested if `parent_ != nullptr`.

### CODE REVIEW

File: cpp/src/arrow/compute/exec/util.cc
Comment (vibhatha) [inline cpp/src/arrow/compute/exec/util.cc:397]: should we move this?

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/extension_set.cc:151]: should we remove this or intended to use?

File: cpp/src/arrow/engine/substrait/function_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/function_test.cc:152]: should we remove this?

Comment (westonpace) [reply cpp/src/arrow/engine/substrait/function_test.cc]: I uncommented this and implemented these cases.

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/extension_set.cc]: I've uncommented this and merged `FunctionMapping` into `ExtensionIdRegistry`.

**Comment by westonpace** [comment]
I moved the `FunctionMapping` into `ExtensionIdRegistry` and reworked a few things.  I think this is ready for review now.  I did try to avoid unnecessary changes but there are a few cleanups in here as well.

@rtpsw this will likely have some (hopefully limited) impact on the UDF work you have ongoing.
@vibhatha I still need to think through what impacts this will have on aggregates.  I'll take a look at that on Monday.

File: cpp/src/arrow/engine/substrait/serde_test.cc
Comment (rtpsw) [inline cpp/src/arrow/engine/substrait/serde_test.cc:709]: I think it would be useful, at least for PyArrow's sake, to expose the string value of this constant (and similar ones) like `default_extension_types_uri()` does.

Comment (westonpace) [reply cpp/src/arrow/engine/substrait/serde_test.cc]: I converted these constants to std::string.

Comment (westonpace) [reply cpp/src/arrow/engine/substrait/serde_test.cc]: I'm not sure if that will make them available to python or not without the functions though.  I'll try that out later.

File: cpp/src/arrow/engine/substrait/serde.h
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/serde.h:122]: I moved the named table provider into the conversion options object being added by @jvanstraten 's PR here: #13537

Hopefully any new options can go into ConversionOptions and that will prevent future creep.

File: cpp/src/arrow/compute/exec/util.cc
Comment (westonpace) [reply cpp/src/arrow/compute/exec/util.cc]: Yes.  Done.

**Comment by westonpace** [comment]
I rebased on top of #13537 so that I could move the named table provider into the conversion options.  I added support for aggregation function converters, at least for the Substrait -> Arrow direction.  I won't finish Arrow -> Substrait for aggregates due  ARROW-17239.

I still need to add tests for Substrait->Arrow aggregates and then I will try and get this PR merged.

**Comment by westonpace** [comment]
I added tests for aggregate functions.  I only added sum/min/max/mean.  The count function will be a bit trickier ([ARROW-17061](https://issues.apache.org/jira/browse/ARROW-17061)).  I think this PR has grown enough however (perhaps a bit too much) and I'd like to just merge in what I have so far and address the rest once the Substrait functions have finalized a bit (or as needed to advance other work).

File: cpp/src/arrow/compute/exec/util.cc
Comment (vibhatha) [inline cpp/src/arrow/compute/exec/util.cc:388]: Not related to this PR, but just a thought, should we move TableSourceNode to utils?

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/expression_internal.cc:84]: nit: Not sure if this is possible, but can we inclue the function.phase() in the message?

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/expression_internal.cc:91]: nit: same here. Just to inform with the tried phase and say it is not supported rather than saying only supporting a particular option only.

**Comment by vibhatha** [comment]
@westonpace General comment on usage of `auto` with `ARROW_ASSIGN_OR_RAISE`. There is a mix of auto and variable type through out the PR. Is there any specific reason? I think there was a recent discussion on minimizing usage of `auto` in certain places where it is hard to deterimine the data type for the readability sake. Is it the case here?

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/ext_test.cc:155]: Should this be `const Id& function_id` instead?

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/ext_test.cc:208]: I know it is not part of your edit, should this be `const util::string_view& name` instead?

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/ext_test.cc:261]: nit: `name1` is this okay? since there is `name2` could it be renamed?

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/extension_set.cc:484]: should we add converter for `function_name` (add function_name in the message) already exists?

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/ext_test.cc:155]: TBH, I don't really know.  Small things we pass by value (e.g. `int`).  A `string_view` is a pair of pointers and generally considered small enough that it is [better to pass by value](https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/).

An `Id` is a pair of `string_view` objects.  Does that mean it is small enough?  I'm not sure what the exact rules are here but the existing usage seems to be treating it as a small value object and so I have tried to continue to do so.

Perhaps @bkietz for a second opinion / further guidelines.

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/ext_test.cc:208]: I linked an article earlier that talks about why it is better to pass `util::string_view` by value.  I find it is helpful to think of `string_view` as just a specialized `char*`.  So I wouldn't do `const char*& name` and I don't do `const util::string_view& name`

File: cpp/src/arrow/compute/exec/util.cc
Comment (westonpace) [reply cpp/src/arrow/compute/exec/util.cc:388]: The `SinkNodeConsumer` objects can be useful to add to the public API because users specify them as part of options.  That's why I had to move this one since the previous location wasn't public and I couldn't use it in tests.

The `TableSourceNode` is a node and those stay in the private API so I don't think we need to move that one.

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:84]: That's a good idea.  I recall someone mentioning something about protobuf descriptors.  I'll see if I can figure out a good way to get this string value.

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/ext_test.cc:261]: I don't understand.  Are you suggesting `name`/`name2` instead of `name1`/`name2`?

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (bkietz) [reply cpp/src/arrow/engine/substrait/ext_test.cc:155]: For anything trivial(ly copyable, destructible, ...) by-reference and by-value should be interchangeable. Using it by value here is more readable so do that. The compiler will have no difficulty removing the trivial copy

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (bkietz) [reply cpp/src/arrow/engine/substrait/ext_test.cc:208]: :+1:

File: cpp/src/arrow/engine/substrait/serde_test.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/serde_test.cc]: I considered adding these functions but, at the end of the day, this is information that the caller should be getting from Substrait and I don't know if Arrow needs to provide them.  I don't want to hold up this PR on this so we still think these functions are needed we can create a follow-up PR and have the discussion there.

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/expression_internal.cc]: I added an `EnumToString` method and used it to print out the invalid invocation / phase that the user provided.  The error looks like this:

```
NotImplemented: Unsupported aggregation invocation 'AGGREGATION_INVOCATION_DISTINCT'.  Only AGGREGATION_INVOCATION_ALL is supported
```

If the user provides a completely invalid value I thought I would get "unknown" but it looks like the google protobuf converter changes it to unspecified so I get:

```
NotImplemented: Unsupported aggregation invocation 'AGGREGATION_INVOCATION_UNSPECIFIED'.  Only AGGREGATION_INVOCATION_ALL is supported
```

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/expression_internal.cc]: Added `EnumToString` here too

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/ext_test.cc:261]: Ping

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/extension_set.cc:488]: I updated this error message to:

```
      return Status::Invalid("Cannot register function converter for Substrait id ",
                             substrait_function_id.ToString(),
                             " because a converter already exists");
```

does that address your concern?

**Comment by westonpace** [comment]
> General comment on usage of auto with ARROW_ASSIGN_OR_RAISE. There is a mix of auto and variable type through out the PR. Is there any specific reason? I think there was a recent discussion on minimizing usage of auto in certain places where it is hard to deterimine the data type for the readability sake. Is it the case here?

The primary places I use `auto` are `make_unique` and for STL iterators.  Looking through this PR I see there are also a number of places where I kept `auto` because I was just moving or refactoring existing code that was already using `auto`.

That being said, there may be a few places where I used `auto` incorrectly.  I'll take another scan.

File: cpp/src/arrow/engine/substrait/serde_test.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/serde_test.cc]: Actually, I had to revert these back to `char *` to make the linter happy.  My thought  on a follow-up PR still holds I think.

**Comment by westonpace** [comment]
@vibhatha / @rtpsw any other concerns?  I'm going to try and merge this tomorrow assuming it passes CI.

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/expression_internal.cc:44]: does `int idx` needs to be `uint32_t idx`?

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/ext_test.cc:262]: Ah no I mean instead it could be an array `name[0]` and `name[1]`. It is not a must here.

File: cpp/src/arrow/engine/substrait/function_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/function_test.cc:65]: should this be instead `Result<std::shared_ptr<Table>>`?

File: cpp/src/arrow/engine/substrait/function_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/function_test.cc:83]: same here, should we use Result?

File: cpp/src/arrow/engine/substrait/function_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/function_test.cc:95]: Should we use `Result` here?

File: cpp/src/arrow/engine/substrait/function_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/function_test.cc:368]: Use `Result`?

File: cpp/src/arrow/engine/substrait/function_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/function_test.cc:355]: use `Result`?

File: cpp/src/arrow/engine/substrait/function_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/function_test.cc:377]: use `Result`?

File: cpp/src/arrow/engine/substrait/relation_internal.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:368]: Thanks for catching this.

File: cpp/src/arrow/engine/substrait/relation_internal.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:394]: Neat!

File: cpp/src/arrow/engine/substrait/serde_test.cc
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/serde_test.cc:738]: did you mean to drop this one?

Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc:738]: `EXPECT_EQ(decoded_add_func.name, "add");`

File: cpp/src/arrow/engine/substrait/test_plan_builder.h
Comment (vibhatha) [inline cpp/src/arrow/engine/substrait/test_plan_builder.h:50]: should we add `docstring` here?

File: python/pyarrow/tests/test_substrait.py
Comment (vibhatha) [inline python/pyarrow/tests/test_substrait.py:152]: Is it wise to assert the supported functions by name rather than checking if it returns a `len > 0` list?

**Review by vibhatha — APPROVED** [review]
@westonpace Looks good to me. I added a few comments for minor changes. 
And thanks working on this and cleaning up the aggregate section with new test cases.

File: cpp/src/arrow/engine/substrait/serde_test.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/serde_test.cc:738]: `DecodeFunction` used to return an `Id` and an Arrow function name.  Returning only an Arrow function name was not sufficient.  Now it only returns an `Id`.  The job was split into `DecodeFunction` (which the extension set provides itself) and `GetSubstraitCallToArrow` (which the extension id registry provides).  So there is no `decoded_add_func.name` any longer.

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/ext_test.cc:262]: Ah, I see your point.  I was just porting over an existing test and I think both are probably ok so I think I'd prefer to leave it as is.

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/expression_internal.cc]: I think I was originally aiming for compatibility with the source, which is a Substrait (i.e. protobuf) index which is signed.  But I agree it looks nicer with `uint32_t` since that matches the destination (and it seems I was inconsistent with casting).  I have updated it.

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/expression_internal.cc]: Yes that’s better.

File: cpp/src/arrow/engine/substrait/ext_test.cc
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/ext_test.cc:262]: Yes. It’s probably fine for this.

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/extension_set.cc:488]: Yes this works.

File: cpp/src/arrow/engine/substrait/serde_test.cc
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc:738]: Got it. I missed that.

**Comment by vibhatha** [comment]
@westonpace there is a link-error

```bash
unity_0_cxx.cxx.obj : error LNK2019: unresolved external symbol "class arrow::Result<class std::shared_ptr<class arrow::compute::ExecPlan> > __cdecl arrow::engine::DeserializePlan(class arrow::Buffer const &,class std::shared_ptr<class arrow::compute::SinkNodeConsumer> const &,class arrow::engine::ExtensionIdRegistry const *,class arrow::engine::ExtensionSet *,struct arrow::engine::ConversionOptions const &)" (?DeserializePlan@engine@arrow@@YA?AV?$Result@V?$shared_ptr@VExecPlan@compute@arrow@@@std@@@2@AEB
```

**Pull request merged**
